staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanagertest/testfieldmanager.go

```diff
@@ -0,0 +1,224 @@
+// objects (unstructured) of the given GVK.
+func NewFakeObjectCreater(gvk schema.GroupVersionKind) runtime.ObjectCreater {
+	return &fakeObjectCreater{gvk: gvk}
+}
+
+type fakeObjectCreater struct {
+	gvk schema.GroupVersionKind
+}
+
+func (f *fakeObjectCreater) New(_ schema.GroupVersionKind) (runtime.Object, error) {
+	u := unstructured.Unstructured{Object: map[string]interface{}{}}
+	u.SetAPIVersion(f.gvk.GroupVersion().String())
+	u.SetKind(f.gvk.Kind)
+	return &u, nil
+}
+
+// TestFieldManager is a FieldManager that can be used in test to
+// simulate the behavior of Server-Side Apply and field tracking. This
+// also has a few methods to get a sense of the state of the object.
+//
+// This TestFieldManager uses a series of "fake" objects to simulate
+// some behavior which come with the limitation that you can only use one version since there is no version conversion logic.
+type TestFieldManager struct {
+	fieldManager *fieldmanager.FieldManager
+	apiVersion   string
+	emptyObj     runtime.Object
+	liveObj      runtime.Object
+}
+
+// NewDefaultTestFieldManager returns a new TestFieldManager built for
+// the given gvk, on the main resource.
+func NewDefaultTestFieldManager(gvk schema.GroupVersionKind) TestFieldManager {
+	return NewTestFieldManager(gvk, "", nil)
+}
+
+// NewTestFieldManager creates a new manager for the given GVK.
+func NewTestFieldManager(gvk schema.GroupVersionKind, subresource string, chainFieldManager func(fieldmanager.Manager) fieldmanager.Manager) TestFieldManager {
+	typeConverter := NewBuiltinTypeConverter()
+	apiVersion := fieldpath.APIVersion(gvk.GroupVersion().String())
+	objectConverter := &fakeObjectConvertor{sameVersionConverter{}, apiVersion}
+	f, err := fieldmanager.NewStructuredMergeManager(
+		typeConverter,
+		objectConverter,
+		&fakeObjectDefaulter{},
+		gvk.GroupVersion(),
+		gvk.GroupVersion(),
+		nil,
+	)
+	if err != nil {
+		panic(err)
+	}
+	live := &unstructured.Unstructured{}
+	live.SetKind(gvk.Kind)
+	live.SetAPIVersion(gvk.GroupVersion().String())
+	f = fieldmanager.NewLastAppliedUpdater(
+		fieldmanager.NewLastAppliedManager(
+			fieldmanager.NewProbabilisticSkipNonAppliedManager(
+				fieldmanager.NewBuildManagerInfoManager(
+					fieldmanager.NewManagedFieldsUpdater(
+						fieldmanager.NewStripMetaManager(f),
+					), gvk.GroupVersion(), subresource,
+				), NewFakeObjectCreater(gvk), gvk, fieldmanager.DefaultTrackOnCreateProbability,
+			), typeConverter, objectConverter, gvk.GroupVersion(),
+		),
+	)
+	if chainFieldManager != nil {
+		f = chainFieldManager(f)
+	}
+	return TestFieldManager{
+		fieldManager: fieldmanager.NewFieldManager(f, subresource),
+		apiVersion:   gvk.GroupVersion().String(),
+		emptyObj:     live,
+		liveObj:      live.DeepCopyObject(),
+	}
+}
+
+// APIVersion of the object that we're tracking.
+func (f *TestFieldManager) APIVersion() string {
+	return f.apiVersion
+}
+
+// Reset removes the currently recorded liveObject.
```

Comment:
Removes or resets to the original state?

Reply:
Reworded